### ISSUE TITLE
LuaGuiElement API call when LuaGuiElement was invalid

### ISSUE THREAD

**Issue by angrypidgeon123** [Gorfiend/Factorio-UnitLocationHotkeys#1]
I changed view from orbit to nauvis and got this error
 
 Running an extensive modlist: https://pastebin.com/mKUSdEBh
 
 I don't know how to check and eliminate this error, please tell or update. Thanks
 
 I tested the same scenario with Location Hotkeys disabled and it doesn't crash
 
 223.657 Script @__RoboTank__/control.lua:163: RoboTank: on_players_changed
 223.657 Error MainLoop.cpp:1404: Exception at tick 23365783: The mod Space Exploration (0.6.121) caused a non-recoverable error.
Please report this error to the mod author.

Error while running event space-exploration::on_gui_click (ID 1)
The mod Unit/Location Hotkeys (1.2.0) caused a non-recoverable error.
Please report this error to the mod author.

Error while running event UnitLocationHotkeys::on_player_changed_surface (ID 53)
LuaGuiElement API call when LuaGuiElement was invalid.
stack traceback:
	[C]: in function '__index'
	__UnitLocationHotkeys__/gui.lua:50: in function 'rebuild_table'
	__UnitLocationHotkeys__/control.lua:469: in function <__UnitLocationHotkeys__/control.lua:465>
stack traceback:
	[C]: in function 'teleport'
	__space-exploration__/scripts/remote-view.lua:756: in function 'start'
	__space-exploration__/scripts/remote-view-gui.lua:606: in function 'callback'
	__space-exploration__/scripts/event.lua:19: in function <__space-exploration__/scripts/event.lua:17>

**Comment by Gorfiend** [comment]
Can't reproduce the crash.

Not really sure what is going on. As far as I can tell from that stack trace, it means somehow the players gui.frame reference became invalid. I'm fairly sure that can't happen with my mod (it's created on player init, then never changed), but it is added to the global "mod_gui.get_frame_flow", so maybe possible some other mod is invalidating it? I think that would mean basically any interaction with my mod would cause a crash, though.

If you provide a save and steps to reproduce, I could look into it more.

**Comment by angrypidgeon123** [comment]
> Can't reproduce the crash.

Surely it needs this huge mod list :D
https://pastebin.com/mKUSdEBh
* any other tweaks to it I posted here, but I doubt they matter: https://www.twitch.tv/angry_pidgeon_123/about

Also a savegame (older but should work, if I don't change the link means I tested and still crashes):
https://file.io/8s7DbmVvMNIc
* it's just before player flight into orbit, which crashes subsurface 1.0.3 mod, unless 1.0.5 is installed which breaks itself functionally :D
* * remove all hd mods (except zerg), baron sludge mod and extended description mod if you don't want a looong load :D I delete them while testing, then restore them from the bin

Someone from discord mod help gave me this possible patch that also failed
`  local gui_data = player_data.gui
  if not(gui_data.frame and gui_data.frame.valid) then
    gui.rebuild_table(player, player_data)
  else
    gui_data.frame.clear()
  end`
  
  like this: https://pastebin.com/2ANpviQj

EDIT: with the savegame provided, and the "fix" mentioned, the crash occurs while flying up with the player in the rocket. The original game crash was when using SE's navigation sattelite view "N" while player in orbit, clicking nauvis to view it

EDIT2: savegame provided crashes the same way as in edit1, with the original code: https://pastebin.com/hriLPGz1

EDIT3: finally, tested with subsurface 1.0.5, it's not it this time :) subsurface 1.0.3 was crashing just when the rocket left the screen on max zoom. The UnitLocationHotkeys crash occurs sometime in the middle of the flight

Game is 1.1.104

**Comment by Gorfiend** [comment]
Well, tried loading that save with all mods, but kept getting startup conflicts for stuff - including space exploration.

I'd have to guess you have specific versions of all those mods, so might not really be feasible for me to test it.

Editing that patch to call `gui.init_player_gui` instead of `gui.rebuild_table` might stop it from crashing, but isn't really a proper fix.

**Comment by angrypidgeon123** [comment]
the save game loading pops up a button called sync mods which takes care of the versions, otherwise I pretty much use the latest. I've compiled the mod pack in the last month. Try loading without mods, then loading the savegame, it should download the mods correctly

I could upload the mods somewhere if it still doesn't work but tomorrow, it's a 2.42GB mod list and is getting late for me

**Comment by Gorfiend** [comment]
I did sync mods with the save, but that only syncs which mods are active - it doesn't sync the specific versions. So I would have downloaded the most recent version of everything, while you may have older versions of some things. I'm guessing if you updated everything, you would see the same errors.

I tried running the save without any mods other than mine, and it immediately crashes on load (the mod migrations cause it to rebuild the gui, which fails with that same LuaGuiElement invalid error). So whatever invalidated the gui frame happened before that save, so I don't think running with all the mods would provide any more info.

I'm not a pro modder, but I'm fairly certain my mod never invalidates the frame, so I have to assume one of the other mods is doing it. I suppose I could change all my code to handle other mods randomly deleting my gui, but that's... ugly. Better for whatever mod is doing it to not invalidate other mods guis. Finding which mod that is, though... don't think I can help with that.

**Comment by heinwessels** [comment]
> I did sync mods with the save, but that only syncs which mods are active - it doesn't sync the specific versions. So I would have downloaded the most recent version of everything, while you may have older versions of some things. I'm guessing if you updated everything, you would see the same errors.

Just a helpful tip, it is possible to sync the exact versions :)

![image](https://github.com/Gorfiend/Factorio-UnitLocationHotkeys/assets/39875036/912852b9-77e1-4ab9-ada0-e80a18f9b0e7)

**Comment by angrypidgeon123** [comment]
Here's the exact mods and settings uploaded
https://www.transferxl.com/download/08jgBfZKB36R5q

EDIT: the hd pack is missing from the upload:
https://mods.factorio.com/explore?query=factoriohd&factorio_version=1.1&sort_attribute=last_updated_at
also baron sludge
https://mods.factorio.com/explore?query=baron&factorio_version=1.1&sort_attribute=last_updated_at
also extended description, because they load very slow and not useful for testing
https://mods.factorio.com/explore?query=extended%20description&factorio_version=1.1&sort_attribute=last_updated_at


Link should last for a week. It's true I don't have the absolute latest versions since I'ld have to test every time a mod uploads, and with this many it'll be a nightmare. Took me 100 hours to go to orbit while testing the modpack...

I guarantee it works as it is on Factorio 1.1.104, well except Location Hotkeys (1.2.0)

Up to you if you fix it. If it conflicts with a popular mod I'ld say that's a problem for your mod popularity. The issue isn't that other mods create problems for yours, it's that if yours doesn't take care of it, the users are unaware of the subtleties of responsibility and will just avoid. e.g. I'm not too happy with ultimate research queue mod, and I even fixed a null error with it, but I got none better atm... same with subsurface mod, there's no underground mod that works better right now

My plan is to have a spider on each planet as a substitute for myself, building stuff, while I personally deliver stuff from the main base and coordinate, so Location Hotkeys fits right into that plan

I'm going to test right now removing Gui Manager, if I say nothing more means it wasn't it :D

I have other subtle problems with these many mods, such as hitting the tile limit which is why I can't install more mods like compaktcircuit (found a substitute) or dectorio

**Comment by angrypidgeon123** [comment]
Fixed. Credit to PenguinSpy https://penguinspy.dev/ from discord factorio mod help channel

this replaces "local gui_data = player_data.gui gui_data.frame.clear()" in gui.lua:50
```

  local gui_data = player_data.gui
  if not(gui_data.frame and gui_data.frame.valid) then
    gui.init_player_gui(player, player_data)
  else
    gui_data.frame.clear()
  end
```
Or at least it doesn't crash :D haven't tested functionally yet. As usual if I say nothing more it should mean it works now

**Comment by Gorfiend** [comment]
> > I did sync mods with the save, but that only syncs which mods are active - it doesn't sync the specific versions. So I would have downloaded the most recent version of everything, while you may have older versions of some things. I'm guessing if you updated everything, you would see the same errors.
> 
> Just a helpful tip, it is possible to sync the exact versions :)

Awesome! That's really useful. So using that had me download different versions of a bunch of mods, and now I'm able to launch and load the save.

...I don't see the crash with my mod, though (just the subsurface crash). Which is strange, because it did crash when I loaded with no other mods... I'll poke at it a bit more, see if I can figure anything else out.


As to the fix - maybe it's an easier change to handle it on my side then I was originally thinking. Will still need to check if there's anything else that might be using an invalid gui.

Even if I do handle it on my side, assuming it's being caused by another mod, it would still be ideal to fix that other mod - it's unlikely my mod is the only victim.

**Comment by angrypidgeon123** [comment]
It's an old save in which mods were added and removed. Maybe something stayed behind causing problems. I heard rumors of a cache too somewhere :D 

Anyway that code bit fixed it and I tested functionally, it works

P.S.

As an ex-programmer my advice is to write complete code meaning allowing only the input you're expecting, and allow no fatal errors but try catch everything https://www.lua.org/pil/8.4.html When I was hired I never produced a bug (because I carefully blamed specification or API :D, point being I foresaw any problems and didn't produced bugs within the parameters I was allowed to function)

Professional testers input positive (what's expected) and negative (what's unexpected) into code. So if your code handles both it will never fail

I never noticed anyone do it though that's a reason to be excellent easily. Only issue with it it's time consuming but once it's incorporated into habit not so much. It's more time consuming to fix bugs than prevent them

**Comment by Gorfiend** [comment]
Pushed up a new version with a fix (slightly different from the one you posted - that could cause it to create two tables in the frame).

After poking at the save a little more, I saw that for some reason the left mod_gui_frame_flow was owned by the mod ChangeMapSettings (normally it's owned by the base Scenario). So if I loaded the save with just my mod and that one, it wouldn't crash - I'm guessing loading without the ChangeMapSettings causes all of the gui pieces it owns to be destroyed, meaning anything in that frame flow (like my frame) would also be destroyed. I took a quick look at the ChangeMapSettings code, but couldn't see how it might take control of that gui piece... so no idea how that situation happened. And still doesn't explain the original crash you were seeing.

**Comment by angrypidgeon123** [comment]
Yes when I loaded it fixed like that, I noticed the gui had a twin view :D I clicked one of them and the other dissapeared, otherwise behaving normal. Usually I don't update until something goes wrong. If it ain't broke don't fix I say :D

EDIT: I eliminated ChangeMapSettings mod a short while ago, but you must've loaded through the save game perhaps because the save game is old, before I did a lot of mod fiddling. That explains some of the crashes you experience then, as since I refined the mod list to the one I uploaded. The old save game retained its own mod list. I tested with the new mod list which overwrites the old save game's on load
